### PR TITLE
Update training controller service to fetch parameters from KV and launch training job

### LINKED ISSUE #1144: [EPIC] Enable AI Services when watchlist is created or updated.
OEP: https://github.com/rancher/opni/blob/main/enhancements/aiops/20230216-enable-gpu-services-training.md

## Issues:
TBD

## Summary: 
Currently, when a user would like to enable AI services, they will go to the Opni Admin Dashboard. First, they are required to enable Logging and once that is done, they will go to the AIOps panel and check the "Enable GPU Services button". When they hit the "Save" button, the GPU services are installed within the Kubernetes cluster. This includes the workload DRAIN service, the training controller service, the GPU Controller service and CPU Inferencing service. This UX can be avoided by simply detecting the availability of a GPU within the cluster and when the user creates or updates the workload log anomaly watchlist to train a Deep Learning model, that is when these GPU services should be installed, rather than through a checkbox button.

## Use case: 
This will remove the "Enable GPU Services" check box and now will install the GPU services when the user decides to update the watchlist for the very first time with workloads. Opni GPU services will automatically come up upon the creation or update of a workload log anomaly watchlist.

## Benefits: 
* Improves the usability of Opni AIOps log anomaly

## Level of Effort: 
* Code implementation: <= 3 days
* Testing and debugging: <= 2 days
* Documentation: 2 days

### LINKED ISSUE #1233: Update AI services to launch training jobs upon startup of service
With the AI services being enabled upon the launching of a training of a Deep Learning model, the training controller service will need to launch a training job upon startup of the service. Additionally, changes will also need to be made to the preprocessing service and inferencing service as well.

### PR DESCRIPTION
**Pull request by AmartC**
This PR implements logic to fetch the workload parameters from Nats KV and will launch a training job of a Deep Learning model if necessary. This is to address the logic within the AIOps gateway plugin which will enable the GPU AI services when a training job is submitted.

### CODE REVIEW

File: training_controller/main.py [hunk 1]
```diff
@@ -124,16 +125,21 @@ def check_training_necessary(updated_workload_parameters: dict):
     return False
 
 
-async def schedule_model_training(workload_parameters: str):
+async def schedule_model_training(workload_parameters: dict):
     global workload_parameters_dict
-    updated_workload_parameters = json.loads(workload_parameters)
-    await update_opensearch(workload_parameters)
+    updated_workload_parameters = copy.deepcopy(workload_parameters)
```
Comment (tybalex) [inline training_controller/main.py]: redundant deep copy? see line 133

File: training_controller/main.py [hunk 2]
```diff
@@ -124,16 +125,21 @@ def check_training_necessary(updated_workload_parameters: dict):
...
     model_training_necessary = check_training_necessary(updated_workload_parameters)
     workload_parameters_dict = copy.deepcopy(updated_workload_parameters)
-    workload_parameter_payload = {"workloads": workload_parameters_dict}
+    workload_parameter_payload = {"workloads": workload_parameters_dict["workloads"]}
 
     if model_training_necessary:
         gpu_service_status = await get_gpu_service_status()
+        num_retries = 0
+        while gpu_service_status != "running" and num_retries < RETRY_LIMIT:
+            asyncio.sleep(30)
```
Comment (tybalex) [inline training_controller/main.py]: await

File: training_controller/main.py [hunk 3]
```diff
@@ -124,16 +125,21 @@ def check_training_necessary(updated_workload_parameters: dict):
...
+            gpu_service_status = await get_gpu_service_status()
+            num_retries += 1
         if gpu_service_status == "running":
```
Comment (tybalex) [inline training_controller/main.py:175]: what happens if the gpu_service is not `running` within 15 retries? would the UI be able to get this info?

File: training_controller/main.py [hunk 4]
```diff
@@ -160,6 +166,7 @@ async def train_model():
         end_ts = int(time.time() * 1000)
         start_ts = end_ts - TRAINING_DATA_INTERVAL
         parentheses_keywords = [f"({x})" for x in ANOMALY_KEYWORDS]
+        workload_parameters = workload_parameters_dict["workloads"]
```
Comment (tybalex) [inline training_controller/main.py:202]: * now that we have natsKV, should probably not using global variable `workload_parameters_dict` anymore, instead just read from natsKV.

File: training_controller/main.py [hunk 5]
```diff
@@ -243,9 +249,16 @@ async def get_latest_workload():
     global workload_parameters_dict
     try:
```
Comment (tybalex) [inline training_controller/main.py:287]: rename this function or add docstrings to match its new activity

File: training_controller/main.py [hunk 6]
```diff
@@ -299,11 +312,11 @@ async def model_status_sub_handler(msg):
...
         reply_subject = msg.reply
-        training_payload = msg.data.decode()
-        if training_payload == "{}":
+        training_payload = json.loads(msg.data.decode())
+        if len(training_payload["workloads"]) == 0:
             global workload_parameters_dict
-            workload_parameters_dict = json.loads(training_payload)
-            await update_opensearch(training_payload)
+            workload_parameters_dict = copy.deepcopy(training_payload)
+            await update_opensearch(training_payload["workloads"])
```
Comment (tybalex) [inline training_controller/main.py]: is it still necessary to push to OpenSearch

**Review by tybalex — CHANGES_REQUESTED** [review]
In general can you please add description of `what does this PR do`

File: training_controller/main.py (hunk 3 above)
Comment (AmartC) [reply training_controller/main.py:175]: No right now, it wouldn't get that information right now. We can look to add logic in future if this step fails.

**Review by tybalex — APPROVED** [review]
lg

**Pull request merged**